### ISSUE TITLE
IndexedLineSet update attributes issue

### ISSUE THREAD

**Issue by ajcyeh** [x3dom/x3dom#582]
When updating attributes of IndexedLineSet, x3dom caused an error:
[GroupMarkerNotSet(crbug.com/242999)!:D064E4BAC57F0000]GL ERROR :GL_INVALID_OPERATION : glDrawElements: attempt to access out of range vertices in attribute 1

This is the intended x3d. It loads correctly.
https://vrmath2.net/VRM2/html/testOLine.html

This html below starts with one line then click on the add line button to update attribute.
https://vrmath2.net/VRM2/html/testOLine1.html
The console will show the "out of range" error and no lines are rendered.
Click on the refresh button to remove the IndexedLineSet and added back in, the lines will be rendered correctly.

This may be related to #580. If multiple colors are also updated, it renders correctly without error.
https://vrmath2.net/VRM2/html/testOLine2.html

If number of colors is less than the number of items in the colorIndex then there is an "out of range" error.
https://vrmath2.net/VRM2/html/testOLine3.html

The x3d spec says "If the greatest index in the colorIndex field is N, there shall be N+1 colours in the Color node." http://www.web3d.org/documents/specifications/19775-1/V3.2/Part01/components/rendering.html#IndexedLineSet

Could x3dom have used the number of items in the colorIndex instead of the "greatest index"?

**Comment by ajcyeh** [comment]
After further investigation I found the filedChanged function seems to miss observing the change of colorIndex in https://github.com/x3dom/x3dom/blob/7d104d9717d12e1670b1ef40f5b83beeb590df83/src/nodes/Rendering/IndexedLineSet.js

Therefore when the colorIndex was updated using setAttribute, x3dom will not update, and instead it is using coordIndex.

This example shows this issue.https://vrmath2.net/VRM2/html/testOLine4.html

The add line button will update the colorIndex to "0 1 0" and the color to "0 0 0,1 0 0,0 1 0". This time there are three colors so it won't cause the out of range error. The add line button produces lines as below from black to red to green, which shows it uses the coordIndex "0 1 2" instead of the colorIndex "0 1 0".
![brg](https://cloud.githubusercontent.com/assets/1114036/11708510/b04fa42e-9f5b-11e5-849c-35d2511e5ff9.png)

Then click on the refresh x3d button, colorIndex will be used to produce the intended outcome.
![brb](https://cloud.githubusercontent.com/assets/1114036/11708549/f5101828-9f5b-11e5-94e4-4073108b70e3.png)

**Comment by tsturm** [comment]
Hi Andy,

can you test it with the latest dev-version.
The problem should be fixed with my last commit.

Regards,
Timo

**Comment by ajcyeh** [comment]
Hi Timo,

Thanks for working on this. I tested the latest dev version in the same example https://vrmath2.net/VRM2/html/testOLine4.html but is getting the following errors in Chrome:

(Click on add line)
Uncaught TypeError: Cannot read property 'x' of undefined
x3dom.registerNodeType.defineClass._buildGeometry @ x3dom.js:3034
x3dom.registerNodeType.defineClass.fieldChanged @ x3dom.js:3051
x3dom.registerNodeType.defineClass.updateField @ x3dom.js:2774
x3dom.setElementAttribute @ x3dom.js:2711
addLine @ testOLine4.html:36
onclick @ testOLine4.html:25

testOLine4.html:1 
[GroupMarkerNotSet(crbug.com/242999)!:D0AC051DFD7F0000]GL ERROR :GL_INVALID_OPERATION : glDrawElements: range out of bounds for buffer

(Click on refresh x3d)
x3dom.js:3034 Uncaught TypeError: Cannot read property 'x' of undefined
x3dom.registerNodeType.defineClass._buildGeometry @ x3dom.js:3034
x3dom.registerNodeType.defineClass.nodeChanged @ x3dom.js:3046
x3dom.NodeNameSpace.setupTree @ x3dom.js:2746
domEventListener.onNodeInserted @ x3dom.js:1100
refresh @ testOLine4.html:43
onclick @ testOLine4.html:26

**Comment by tsturm** [comment]
Hi Andy, thanks for testing.

It seems that i have rearranged the setAttribute calls in my test file at some point.
I first set the new coordinates and colors and only after that the new indices.

The following addLine function works for me but I'll try that your version also works.

```
function addLine() {
    var s = document.getElementById("obj_0_shp");
    var l = document.getElementById("obj_0_line");
    var coord = document.getElementById("obj_0_coor");
    var color = document.getElementById("obj_0_color");
    coord.setAttribute("point", "0 0 0, 0 1 0, 1 1 0");
    color.setAttribute("color", "0 0 0, 1 0 0");		
    l.setAttribute("coordIndex", "0 1 2");
    l.setAttribute("colorIndex", "0 1 0");
}
```

**Comment by tsturm** [comment]
Ok, the new fix is there. Can you try it again with the latest dev-version?

**Comment by ajcyeh** [comment]
Hi Timo,

Thanks a lot. This fix works for IndexedLineSet.
However, I just found that IndexedFaceSet has the same issue. I will open another issue to explain.

Andy :-)
